core/storage/trie/polkadot_trie/polkadot_trie_cursor_impl.cpp

```diff
@@ -101,159 +110,149 @@ namespace kagome::storage::trie {
+          default:
+            return Error::INVALID_NODE_TYPE;
+        }
+      }
+      // if the left part of the sought key exceeds the current node's key, but
+      // its prefix is equal to the key, then we proceed to a child node that
+      // starts with a nibble that is greater of equal to the first nibble of
+      // the left part (if there is no such child, proceed to the next case)
+      bool longer = (part_equal and left_mismatch != left_nibbles.end()
+                     and current_mismatch == current->key_nibbles.end());
+      if (longer) {
+        switch (current->getTrieType()) {
+          case NodeType::BranchEmptyValue:
+          case NodeType::BranchWithValue: {
+            auto current_as_branch =
+                std::dynamic_pointer_cast<BranchNode>(current);
+            OUTCOME_TRY(child_idx,
+                        getChildWithMinIdx(current_as_branch, *left_mismatch));
+            if (child_idx != -1) {
+              last_visited_child_.emplace_back(current_as_branch, child_idx);
+              OUTCOME_TRY(new_current,
+                          trie_.retrieveChild(current_as_branch, child_idx));
+              left_nibbles =
+                  left_nibbles.subspan(current->key_nibbles.size() + 1);
+              current = new_current;
+              continue;
+            }
+            break;  // go to case3
           }
-          followed_path.emplace_back(current_as_branch,
-                                     left_nibbles[common_length]);
-          current = n;
-          left_nibbles = left_nibbles.subspan(common_length + 1);
-          break;
+          case NodeType::Leaf: {
+            break;  // go to case3
+          }
+          default:
+            return Error::INVALID_NODE_TYPE;
         }
-        case NodeType::Leaf:
-          left_nibbles = left_nibbles.subspan(common_length);
-          goto END;
-
-        case NodeType::Special:
-          return Error::INVALID_NODE_TYPE;
+      }
+      // if the left part of the key is longer than the current of
+      // lexicographically greater than the current, we must return to its
+      // parent and find a child greater than the current one
+      bool longerOrBigger =
+          longer or (not part_equal and *left_mismatch > *current_mismatch);
+      if (longerOrBigger) {
+        while (not last_visited_child_.empty()) {
+          auto [parent, idx] = last_visited_child_.back();
+          last_visited_child_.pop_back();
+          OUTCOME_TRY(child_idx, getChildWithMinIdx(parent, idx + 1));
+          if (child_idx != -1) {
+            OUTCOME_TRY(child, trie_.retrieveChild(parent, child_idx));
+            OUTCOME_TRY(node, seekNodeWithValue(child));
+            current_ = node;
+            return outcome::success();
+          }
+        }
+        current_ = nullptr;
+        return outcome::success();
       }
     }
-  END:
-    return followed_path;
+    BOOST_ASSERT_MSG(false, "Unreachable");
```

Comment:
Look at **UNREACHABLE** macro defined in **core/macro/unreachable.hpp**

Reply:
Oh, thank you, forgot about it. Actually I think that it would be better to refactor the loop condition so that this line is not unreachable, but I can't see any imminent problem with this, nor any good solution, so I'll leave it as it is for now.